### PR TITLE
Don't transfer layout outputs to java for unset edges

### PR DESCRIPTION
**Pull request by woehrl01**
See facebook/yoga#483. We should not transfer the layout if the layout didn't change. (using ```hasNewLayout```). This also changes that the lock on the java node is only aquired if needed, and it holds the lock only for the time the values are set and not for the time all it's children are set.

### CODE REVIEW

**Comment by emilsjolander** [comment]
I'm not sure we can directly use hasNewValue as it is not ever set to false in the yoga code. Yoga.c definetly needs some changes as well. There are also times when the padding of a node could potentially have change but not the width which me may or may not want to handle.

**Comment by woehrl01** [comment]
We could do the following: We use ```hasNewLayout``` just for jni, and keep the value in java as a simple property (no jni callback) and set this value from ```YGTransferLayoutOutputsRecursive``` as well. What do you think?

**Comment by emilsjolander** [comment]
Good idea! I think that should work. React-native depends quite heavily on that value so while Yoga's tests will pass I'll need to run through some internal RN apps to check that nothing unexpected breaks. But I think this idea will probably work.

**Comment by woehrl01** [comment]
I also added that we reuse the ```mHasSetMargin/Padding/Border``` flags. We make 3 calls more than usual, but as I think that not all node have all of these values set (if at all) we at safe 9 jni calls in the best case.

**Comment by woehrl01** [comment]
Further reduces the jni calls by 2 for the edges. By moving from bool to an int flag which stores all set edges. We could even provide a method which does something like ```TouchFromJni()``` which sets the ```hasNewLayout``` flag and returns all the flags in a single call.

File: java/com/facebook/yoga/YogaNode.java [hunk 1]
```diff
@@ -42,9 +42,13 @@ public static void setLogger(YogaLogger logger) {
   private long mNativePointer;
   private Object mData;
 
-  private boolean mHasSetPadding = false;
-  private boolean mHasSetMargin = false;
-  private boolean mHasSetBorder = false;
+  private static int MARGIN = 1;
```
Comment (astreet) [inline java/com/facebook/yoga/YogaNode.java]: make these final

File: java/jni/YGJNI.cpp [hunk 2]
```diff
@@ -24,54 +24,74 @@ static void YGTransferLayoutDirection(YGNodeRef node, alias_ref<jobject> javaNod
...
+
+      static auto borderLeftField = obj->getClass()->getField<jfloat>("mBorderLeft");
+      static auto borderTopField = obj->getClass()->getField<jfloat>("mBorderTop");
+      static auto borderRightField = obj->getClass()->getField<jfloat>("mBorderRight");
+      static auto borderBottomField = obj->getClass()->getField<jfloat>("mBorderBottom");
+
+      static auto hasNewLayoutField = obj->getClass()->getField<jboolean>("mHasNewLayout");
+      static auto edgeSetFlagField = obj->getClass()->getField<jint>("mEdgeSetFlag");
+
+      const int MARGIN = 1;
```
Comment (astreet) [inline java/jni/YGJNI.cpp]: comment that these need to be kept in sync with YogaNode.java (or pull them off that class)

**Comment by astreet** [comment]
Nice, I like this: can we do the hasNewLayout change in a separate PR though? I ask in case something breaks here we can debug easier and revert a more isolated diff.

**Comment by woehrl01** [comment]
@astreet Sure, will put up a separte PR and apply the changes tonight. Thanks!

**Comment by woehrl01** [comment]
On the other hand we could provide a ```ExperimentalFeatureOptimizeJniTransfer``` which we could enable/disable if anything breaks.

**Comment by astreet** [comment]
Thanks! I'm fine with this not being an experimental flag :)

**Comment by emilsjolander** [comment]
I tried landing a much smaller scope version of this earlier which broke a bunch of stuff. I'm going to try to land that first still before landing this but I really like what you have done here. I'll ping you when it is in.

**Comment by emilsjolander** [comment]
Can you rebase this? 👍

File: java/jni/YGJNI.cpp [hunk 3]
```diff
@@ -46,36 +46,51 @@ static void YGTransferLayoutOutputsRecursive(YGNodeRef root) {
...
       YGTransferLayoutDirection(root, obj);
       YGNodeSetHasNewLayout(root, false);
-      for (uint32_t i = 0; i < YGNodeGetChildCount(root); i++) {
-        YGTransferLayoutOutputsRecursive(YGNodeGetChild(root, i));
-      }
     } else {
       YGLog(YGLogLevelError, "Java YGNode was GCed during layout calculation\n");
+      return;
+    }
+    for (uint32_t i = 0; i < YGNodeGetChildCount(root); i++) {
```
Comment (emilsjolander) [inline java/jni/YGJNI.cpp]: This feels like a rebase error

Comment (woehrl01) [inline java/jni/YGJNI.cpp]: This actually isn't one, I think we don't need to lock each object for all it's children. Thats the reason I put the iteration outside the lock. But maybe this should be a separate PR?

Comment (emilsjolander) [inline java/jni/YGJNI.cpp]: I got confused by which if statement it was inside of 👍  but yeah if you could make that a separate PR that would be awesome!

Comment (woehrl01) [inline java/jni/YGJNI.cpp]: of course 👍
